File: SignalServiceKit/src/SSKEnvironment.m
//
// Copyright 2017 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "SSKEnvironment.h"
#import "AppContext.h"
#import "ProfileManagerProtocol.h"
#import "TSAccountManager.h"
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

NSNotificationName const WarmCachesNotification = @"WarmCachesNotification";

static SSKEnvironment *sharedSSKEnvironment;

@interface SSKEnvironment ()

@property (nonatomic) id<ContactsManagerProtocol> contactsManagerRef;
@property (nonatomic) MessageSender *messageSenderRef;
@property (nonatomic) id<ProfileManagerProtocol> profileManagerRef;
@property (nonatomic) NetworkManager *networkManagerRef;
@property (nonatomic) OWSMessageManager *messageManagerRef;
@property (nonatomic) BlockingManager *blockingManagerRef;
@property (nonatomic) OWSIdentityManager *identityManagerRef;
@property (nonatomic) id<OWSUDManager> udManagerRef;
@property (nonatomic) OWSMessageDecrypter *messageDecrypterRef;
@property (nonatomic) GroupsV2MessageProcessor *groupsV2MessageProcessorRef;
@property (nonatomic) SocketManager *socketManagerRef;
@property (nonatomic) TSAccountManager *tsAccountManagerRef;
@property (nonatomic) OWS2FAManager *ows2FAManagerRef;
@property (nonatomic) OWSDisappearingMessagesJob *disappearingMessagesJobRef;
@property (nonatomic) OWSReceiptManager *receiptManagerRef;
@property (nonatomic) OWSOutgoingReceiptManager *outgoingReceiptManagerRef;
@property (nonatomic) id<SyncManagerProtocol> syncManagerRef;
@property (nonatomic) id<SSKReachabilityManager> reachabilityManagerRef;
@property (nonatomic) id<OWSTypingIndicators> typingIndicatorsRef;
@property (nonatomic) OWSAttachmentDownloads *attachmentDownloadsRef;
@property (nonatomic) SignalServiceAddressCache *signalServiceAddressCacheRef;
@property (nonatomic) id<OWSSignalServiceProtocol> signalServiceeRef;
@property (nonatomic) StickerManager *stickerManagerRef;
@property (nonatomic) SDSDatabaseStorage *databaseStorageRef;
@property (nonatomic) StorageCoordinator *storageCoordinatorRef;
@property (nonatomic) SSKPreferences *sskPreferencesRef;
@property (nonatomic) id<GroupsV2> groupsV2Ref;
@property (nonatomic) id<GroupV2Updates> groupV2UpdatesRef;
@property (nonatomic) MessageFetcherJob *messageFetcherJobRef;
@property (nonatomic) BulkProfileFetch *bulkProfileFetchRef;
@property (nonatomic) BulkUUIDLookup *bulkUUIDLookupRef;
@property (nonatomic) id<VersionedProfiles> versionedProfilesRef;
@property (nonatomic) ModelReadCaches *modelReadCachesRef;
@property (nonatomic) EarlyMessageManager *earlyMessageManagerRef;
@property (nonatomic) OWSMessagePipelineSupervisor *messagePipelineSupervisorRef;
@property (nonatomic) AppExpiry *appExpiryRef;
@property (nonatomic) id<PaymentsHelper> paymentsHelperRef;
@property (nonatomic) id<PaymentsCurrencies> paymentsCurrenciesRef;
@property (nonatomic) id<PaymentsEvents> paymentsEventsRef;
@property (nonatomic) id<MobileCoinHelper> mobileCoinHelperRef;
@property (nonatomic) SpamChallengeResolver *spamChallengeResolverRef;
@property (nonatomic) SenderKeyStore *senderKeyStoreRef;
@property (nonatomic) PhoneNumberUtil *phoneNumberUtilRef;
@property (nonatomic) id<WebSocketFactory> webSocketFactoryRef;
@property (nonatomic) ChangePhoneNumber *changePhoneNumberRef;
@property (nonatomic) id<SubscriptionManagerProtocol> subscriptionManagerRef;
@property (nonatomic) id<SystemStoryManagerProtocolObjc> systemStoryManagerRef;

@end

#pragma mark -

@implementation SSKEnvironment {
    SignalProtocolStore *_aciSignalProtocolStoreRef;
    SignalProtocolStore *_pniSignalProtocolStoreRef;
}

@synthesize callMessageHandlerRef = _callMessageHandlerRef;
@synthesize notificationsManagerRef = _notificationsManagerRef;

- (instancetype)initWithContactsManager:(id<ContactsManagerProtocol>)contactsManager
                     linkPreviewManager:(OWSLinkPreviewManager *)linkPreviewManager
                          messageSender:(MessageSender *)messageSender
                  messageSenderJobQueue:(MessageSenderJobQueue *)messageSenderJobQueue
                 pendingReceiptRecorder:(id<PendingReceiptRecorder>)pendingReceiptRecorder
                         profileManager:(id<ProfileManagerProtocol>)profileManager
                         networkManager:(NetworkManager *)networkManager
                         messageManager:(OWSMessageManager *)messageManager
                        blockingManager:(BlockingManager *)blockingManager
                        identityManager:(OWSIdentityManager *)identityManager
                    remoteConfigManager:(id<RemoteConfigManager>)remoteConfigManager
                 aciSignalProtocolStore:(SignalProtocolStore *)aciSignalProtocolStore
                 pniSignalProtocolStore:(SignalProtocolStore *)pniSignalProtocolStore
                              udManager:(id<OWSUDManager>)udManager
                       messageDecrypter:(OWSMessageDecrypter *)messageDecrypter
               groupsV2MessageProcessor:(GroupsV2MessageProcessor *)groupsV2MessageProcessor
                          socketManager:(SocketManager *)socketManager
                       tsAccountManager:(TSAccountManager *)tsAccountManager
                          ows2FAManager:(OWS2FAManager *)ows2FAManager
                disappearingMessagesJob:(OWSDisappearingMessagesJob *)disappearingMessagesJob
                         receiptManager:(OWSReceiptManager *)receiptManager
                 outgoingReceiptManager:(OWSOutgoingReceiptManager *)outgoingReceiptManager
                    reachabilityManager:(id<SSKReachabilityManager>)reachabilityManager
                            syncManager:(id<SyncManagerProtocol>)syncManager
                       typingIndicators:(id<OWSTypingIndicators>)typingIndicators
                    attachmentDownloads:(OWSAttachmentDownloads *)attachmentDownloads
                         stickerManager:(StickerManager *)stickerManager
                        databaseStorage:(SDSDatabaseStorage *)databaseStorage
              signalServiceAddressCache:(SignalServiceAddressCache *)signalServiceAddressCache
                          signalService:(id<OWSSignalServiceProtocol>)signalService
                   accountServiceClient:(AccountServiceClient *)accountServiceClient
                  storageServiceManager:(id<StorageServiceManagerProtocol>)storageServiceManager
                     storageCoordinator:(StorageCoordinator *)storageCoordinator
                         sskPreferences:(SSKPreferences *)sskPreferences
                               groupsV2:(id<GroupsV2>)groupsV2
                         groupV2Updates:(id<GroupV2Updates>)groupV2Updates
                      messageFetcherJob:(MessageFetcherJob *)messageFetcherJob
                       bulkProfileFetch:(BulkProfileFetch *)bulkProfileFetch
                         bulkUUIDLookup:(BulkUUIDLookup *)bulkUUIDLookup
                      versionedProfiles:(id<VersionedProfiles>)versionedProfiles
                        modelReadCaches:(ModelReadCaches *)modelReadCaches
                    earlyMessageManager:(EarlyMessageManager *)earlyMessageManager
              messagePipelineSupervisor:(OWSMessagePipelineSupervisor *)messagePipelineSupervisor
                              appExpiry:(AppExpiry *)appExpiry
                       messageProcessor:(MessageProcessor *)messageProcessor
                         paymentsHelper:(id<PaymentsHelper>)paymentsHelper
                     paymentsCurrencies:(id<PaymentsCurrencies>)paymentsCurrencies
                         paymentsEvents:(id<PaymentsEvents>)paymentsEvents
                       mobileCoinHelper:(id<MobileCoinHelper>)mobileCoinHelper
                  spamChallengeResolver:(SpamChallengeResolver *)spamResolver
                         senderKeyStore:(SenderKeyStore *)senderKeyStore
                        phoneNumberUtil:(PhoneNumberUtil *)phoneNumberUtil
                       webSocketFactory:(id<WebSocketFactory>)webSocketFactory
                      changePhoneNumber:(ChangePhoneNumber *)changePhoneNumber
                    subscriptionManager:(id<SubscriptionManagerProtocol>)subscriptionManager
                     systemStoryManager:(id<SystemStoryManagerProtocolObjc>)systemStoryManager
{
    self = [super init];
    if (!self) {
        return self;
    }
    
    _contactsManagerRef = contactsManager;
    _linkPreviewManagerRef = linkPreviewManager;
    _messageSenderRef = messageSender;
    _messageSenderJobQueueRef = messageSenderJobQueue;
    _pendingReceiptRecorderRef = pendingReceiptRecorder;
    _profileManagerRef = profileManager;
    _networkManagerRef = networkManager;
    _messageManagerRef = messageManager;
    _blockingManagerRef = blockingManager;
    _identityManagerRef = identityManager;
    _remoteConfigManagerRef = remoteConfigManager;
    _aciSignalProtocolStoreRef = aciSignalProtocolStore;
    _pniSignalProtocolStoreRef = pniSignalProtocolStore;
    _udManagerRef = udManager;
    _messageDecrypterRef = messageDecrypter;
    _groupsV2MessageProcessorRef = groupsV2MessageProcessor;
    _socketManagerRef = socketManager;
    _tsAccountManagerRef = tsAccountManager;
    _ows2FAManagerRef = ows2FAManager;
    _disappearingMessagesJobRef = disappearingMessagesJob;
    _receiptManagerRef = receiptManager;
    _outgoingReceiptManagerRef = outgoingReceiptManager;
    _syncManagerRef = syncManager;
    _reachabilityManagerRef = reachabilityManager;
    _typingIndicatorsRef = typingIndicators;
    _attachmentDownloadsRef = attachmentDownloads;
    _stickerManagerRef = stickerManager;
    _databaseStorageRef = databaseStorage;
    _signalServiceAddressCacheRef = signalServiceAddressCache;
    _signalServiceRef = signalService;
    _accountServiceClientRef = accountServiceClient;
    _storageServiceManagerRef = storageServiceManager;
    _storageCoordinatorRef = storageCoordinator;
    _sskPreferencesRef = sskPreferences;
    _groupsV2Ref = groupsV2;
    _groupV2UpdatesRef = groupV2Updates;
    _messageFetcherJobRef = messageFetcherJob;
    _bulkProfileFetchRef = bulkProfileFetch;
    _versionedProfilesRef = versionedProfiles;
    _bulkUUIDLookupRef = bulkUUIDLookup;
    _modelReadCachesRef = modelReadCaches;
    _earlyMessageManagerRef = earlyMessageManager;
    _messagePipelineSupervisorRef = messagePipelineSupervisor;
    _appExpiryRef = appExpiry;
    _messageProcessorRef = messageProcessor;
    _paymentsHelperRef = paymentsHelper;
    _paymentsCurrenciesRef = paymentsCurrencies;
    _paymentsEventsRef = paymentsEvents;
    _mobileCoinHelperRef = mobileCoinHelper;
    _spamChallengeResolverRef = spamResolver;
    _senderKeyStoreRef = senderKeyStore;
    _phoneNumberUtilRef = phoneNumberUtil;
    _webSocketFactoryRef = webSocketFactory;
    _changePhoneNumberRef = changePhoneNumber;
    _subscriptionManagerRef = subscriptionManager;
    _systemStoryManagerRef = systemStoryManager;

    return self;
}

+ (instancetype)shared
{
    OWSAssertDebug(sharedSSKEnvironment);
    
    return sharedSSKEnvironment;
}

+ (void)setShared:(SSKEnvironment *)env
{
    OWSAssertDebug(env);
    OWSAssertDebug(!sharedSSKEnvironment || CurrentAppContext().isRunningTests);
    
    sharedSSKEnvironment = env;
}

+ (BOOL)hasShared
{
    return sharedSSKEnvironment != nil;
}

#pragma mark - Mutable Accessors

- (nullable id<OWSCallMessageHandler>)callMessageHandlerRef
{
    @synchronized(self) {
        OWSAssertDebug(_callMessageHandlerRef);
        
        return _callMessageHandlerRef;
    }
}

- (void)setCallMessageHandlerRef:(nullable id<OWSCallMessageHandler>)callMessageHandlerRef
{
    @synchronized(self) {
        OWSAssertDebug(callMessageHandlerRef);
        OWSAssertDebug(!_callMessageHandlerRef);
        
        _callMessageHandlerRef = callMessageHandlerRef;
    }
}

- (nullable id<NotificationsProtocol>)notificationsManagerRef
{
    @synchronized(self) {
        OWSAssertDebug(_notificationsManagerRef);
        
        return _notificationsManagerRef;
    }
}

- (void)setNotificationsManagerRef:(nullable id<NotificationsProtocol>)notificationsManagerRef
{
    @synchronized(self) {
        OWSAssertDebug(notificationsManagerRef);
        OWSAssertDebug(!_notificationsManagerRef);
        
        _notificationsManagerRef = notificationsManagerRef;
    }
}

- (SignalProtocolStore *)signalProtocolStoreRefForIdentity:(OWSIdentity)identity
{
    switch (identity) {
        case OWSIdentityACI:
            return _aciSignalProtocolStoreRef;
        case OWSIdentityPNI:
            return _pniSignalProtocolStoreRef;
    }
}

- (BOOL)isComplete
{
    return (self.callMessageHandler != nil && self.notificationsManager != nil);
}

- (void)warmCaches
{
    NSArray *specs = @[
        @"signalProxy",
        ^{ [SignalProxy warmCaches]; },
        @"tsAccountManager",
        ^{ [self.tsAccountManager warmCaches]; },
        @"signalServiceAddressCache",
        ^{ [self.signalServiceAddressCache warmCaches]; },
        @"signalService",
        ^{ [self.signalService warmCaches]; },
        @"remoteConfigManager",
        ^{ [self.remoteConfigManager warmCaches]; },
        @"udManager",
        ^{ [self.udManager warmCaches]; },
        @"blockingManager",
        ^{ [self.blockingManager warmCaches]; },
        @"profileManager",
        ^{ [self.profileManager warmCaches]; },
        @"receiptManager",
        ^{ [self.receiptManager prepareCachedValues]; },
        @"OWSKeyBackupService",
        ^{ [OWSKeyBackupService warmCaches]; },
        @"PinnedThreadManager",
        ^{ [PinnedThreadManager warmCaches]; },
        @"typingIndicatorsImpl",
        ^{ [self.typingIndicatorsImpl warmCaches]; },
        @"paymentsHelper",
        ^{ [self.paymentsHelper warmCaches]; },
        @"paymentsCurrencies",
        ^{ [self.paymentsCurrencies warmCaches]; },
        @"storyManager",
        ^{ [StoryManager setup]; }
    ];

    for (int i = 0; i < specs.count / 2; i++) {
        [InstrumentsMonitor measureWithCategory:@"appstart"
                                         parent:@"caches"
                                           name:[specs objectAtIndex:2 * i]
                                          block:[specs objectAtIndex:2 * i + 1]];
    }
    [NSNotificationCenter.defaultCenter postNotificationName:WarmCachesNotification object:nil];
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/src/SSKEnvironment.swift
//
// Copyright 2023 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation

public class SSKEnvironment: NSObject {

    private static var _shared: SSKEnvironment?

    public static var hasShared: Bool { _shared != nil }

    @objc
    public static var shared: SSKEnvironment { _shared! }

    public static func setShared(_ env: SSKEnvironment, isRunningTests: Bool) {
        owsAssert(_shared == nil || isRunningTests)
        _shared = env
    }

    #if TESTABLE_BUILD
    private(set) public var contactsManagerRef: ContactsManagerProtocol
    private(set) public var messageSenderRef: MessageSender
    private(set) public var networkManagerRef: NetworkManager
    private(set) public var paymentsHelperRef: PaymentsHelperSwift
    private(set) public var groupsV2Ref: GroupsV2Swift
    #else
    public let contactsManagerRef: ContactsManagerProtocol
    public let messageSenderRef: MessageSender
    public let networkManagerRef: NetworkManager
    public let paymentsHelperRef: PaymentsHelperSwift
    public let tsAccountManagerRef: TSAccountManager
    public let groupsV2Ref: GroupsV2Swift
    #endif

    public let linkPreviewManagerRef: OWSLinkPreviewManager
    public let pendingReceiptRecorderRef: PendingReceiptRecorder
    public let profileManagerRef: ProfileManagerProtocol
    public let messageManagerRef: OWSMessageManager
    public let blockingManagerRef: BlockingManager
    public let remoteConfigManagerRef: RemoteConfigManager
    public let udManagerRef: OWSUDManager
    public let messageDecrypterRef: OWSMessageDecrypter
    public let groupsV2MessageProcessorRef: GroupsV2MessageProcessor
    public let socketManagerRef: SocketManager
    public let ows2FAManagerRef: OWS2FAManager
    public let disappearingMessagesJobRef: OWSDisappearingMessagesJob
    public let receiptManagerRef: OWSReceiptManager
    public let outgoingReceiptManagerRef: OWSOutgoingReceiptManager
    public let reachabilityManagerRef: SSKReachabilityManager
    public let syncManagerRef: SyncManagerProtocol
    public let typingIndicatorsRef: TypingIndicators
    public let attachmentDownloadsRef: OWSAttachmentDownloads
    public let stickerManagerRef: StickerManager
    public let databaseStorageRef: SDSDatabaseStorage
    public let signalServiceAddressCacheRef: SignalServiceAddressCache
    public let signalServiceRef: OWSSignalServiceProtocol
    public let accountServiceClientRef: AccountServiceClient
    public let storageServiceManagerRef: StorageServiceManager
    public let storageCoordinatorRef: StorageCoordinator
    public let sskPreferencesRef: SSKPreferences
    public let groupV2UpdatesRef: GroupV2UpdatesSwift
    public let messageFetcherJobRef: MessageFetcherJob
    public let bulkProfileFetchRef: BulkProfileFetch
    public let versionedProfilesRef: VersionedProfilesSwift
    public let modelReadCachesRef: ModelReadCaches
    public let earlyMessageManagerRef: EarlyMessageManager
    public let messagePipelineSupervisorRef: MessagePipelineSupervisor
    public let messageProcessorRef: MessageProcessor
    public let paymentsCurrenciesRef: PaymentsCurrenciesSwift
    public let paymentsEventsRef: PaymentsEvents
    public let mobileCoinHelperRef: MobileCoinHelper
    public let spamChallengeResolverRef: SpamChallengeResolver
    public let senderKeyStoreRef: SenderKeyStore
    public let phoneNumberUtilRef: PhoneNumberUtil
    public let webSocketFactoryRef: WebSocketFactory
    public let legacyChangePhoneNumberRef: LegacyChangePhoneNumber
    public let subscriptionManagerRef: SubscriptionManager
    public let systemStoryManagerRef: SystemStoryManagerProtocol
    public let remoteMegaphoneFetcherRef: RemoteMegaphoneFetcher
    public let sskJobQueuesRef: SSKJobQueues
    public let contactDiscoveryManagerRef: ContactDiscoveryManager
    public let callMessageHandlerRef: OWSCallMessageHandler
    public let notificationsManagerRef: NotificationsProtocol
    public let messageSendLogRef: MessageSendLog

    private let appExpiryRef: AppExpiry
    private let aciSignalProtocolStoreRef: SignalProtocolStore
    private let pniSignalProtocolStoreRef: SignalProtocolStore

    public init(
        contactsManager: ContactsManagerProtocol,
        linkPreviewManager: OWSLinkPreviewManager,
        messageSender: MessageSender,
        pendingReceiptRecorder: PendingReceiptRecorder,
        profileManager: ProfileManagerProtocol,
        networkManager: NetworkManager,
        messageManager: OWSMessageManager,
        blockingManager: BlockingManager,
        remoteConfigManager: RemoteConfigManager,
        aciSignalProtocolStore: SignalProtocolStore,
        pniSignalProtocolStore: SignalProtocolStore,
        udManager: OWSUDManager,
        messageDecrypter: OWSMessageDecrypter,
        groupsV2MessageProcessor: GroupsV2MessageProcessor,
        socketManager: SocketManager,
        ows2FAManager: OWS2FAManager,
        disappearingMessagesJob: OWSDisappearingMessagesJob,
        receiptManager: OWSReceiptManager,
        outgoingReceiptManager: OWSOutgoingReceiptManager,
        reachabilityManager: SSKReachabilityManager,
        syncManager: SyncManagerProtocol,
        typingIndicators: TypingIndicators,
        attachmentDownloads: OWSAttachmentDownloads,
        stickerManager: StickerManager,
        databaseStorage: SDSDatabaseStorage,
        signalServiceAddressCache: SignalServiceAddressCache,
        signalService: OWSSignalServiceProtocol,
        accountServiceClient: AccountServiceClient,
        storageServiceManager: StorageServiceManager,
        storageCoordinator: StorageCoordinator,
        sskPreferences: SSKPreferences,
        groupsV2: GroupsV2Swift,
        groupV2Updates: GroupV2UpdatesSwift,
        messageFetcherJob: MessageFetcherJob,
        bulkProfileFetch: BulkProfileFetch,
        versionedProfiles: VersionedProfilesSwift,
        modelReadCaches: ModelReadCaches,
        earlyMessageManager: EarlyMessageManager,
        messagePipelineSupervisor: MessagePipelineSupervisor,
        appExpiry: AppExpiry,
        messageProcessor: MessageProcessor,
        paymentsHelper: PaymentsHelperSwift,
        paymentsCurrencies: PaymentsCurrenciesSwift,
        paymentsEvents: PaymentsEvents,
        mobileCoinHelper: MobileCoinHelper,
        spamChallengeResolver: SpamChallengeResolver,
        senderKeyStore: SenderKeyStore,
        phoneNumberUtil: PhoneNumberUtil,
        webSocketFactory: WebSocketFactory,
        legacyChangePhoneNumber: LegacyChangePhoneNumber,
        subscriptionManager: SubscriptionManager,
        systemStoryManager: SystemStoryManagerProtocol,
        remoteMegaphoneFetcher: RemoteMegaphoneFetcher,
        sskJobQueues: SSKJobQueues,
        contactDiscoveryManager: ContactDiscoveryManager,
        callMessageHandler: OWSCallMessageHandler,
        notificationsManager: NotificationsProtocol,
        messageSendLog: MessageSendLog
    ) {
        self.contactsManagerRef = contactsManager
        self.linkPreviewManagerRef = linkPreviewManager
        self.messageSenderRef = messageSender
        self.pendingReceiptRecorderRef = pendingReceiptRecorder
        self.profileManagerRef = profileManager
        self.networkManagerRef = networkManager
        self.messageManagerRef = messageManager
        self.blockingManagerRef = blockingManager
        self.remoteConfigManagerRef = remoteConfigManager
        self.aciSignalProtocolStoreRef = aciSignalProtocolStore
        self.pniSignalProtocolStoreRef = pniSignalProtocolStore
        self.udManagerRef = udManager
        self.messageDecrypterRef = messageDecrypter
        self.groupsV2MessageProcessorRef = groupsV2MessageProcessor
        self.socketManagerRef = socketManager
        self.ows2FAManagerRef = ows2FAManager
        self.disappearingMessagesJobRef = disappearingMessagesJob
        self.receiptManagerRef = receiptManager
        self.outgoingReceiptManagerRef = outgoingReceiptManager
        self.syncManagerRef = syncManager
        self.reachabilityManagerRef = reachabilityManager
        self.typingIndicatorsRef = typingIndicators
        self.attachmentDownloadsRef = attachmentDownloads
        self.stickerManagerRef = stickerManager
        self.databaseStorageRef = databaseStorage
        self.signalServiceAddressCacheRef = signalServiceAddressCache
        self.signalServiceRef = signalService
        self.accountServiceClientRef = accountServiceClient
        self.storageServiceManagerRef = storageServiceManager
        self.storageCoordinatorRef = storageCoordinator
        self.sskPreferencesRef = sskPreferences
        self.groupsV2Ref = groupsV2
        self.groupV2UpdatesRef = groupV2Updates
        self.messageFetcherJobRef = messageFetcherJob
        self.bulkProfileFetchRef = bulkProfileFetch
        self.versionedProfilesRef = versionedProfiles
        self.modelReadCachesRef = modelReadCaches
        self.earlyMessageManagerRef = earlyMessageManager
        self.messagePipelineSupervisorRef = messagePipelineSupervisor
        self.appExpiryRef = appExpiry
        self.messageProcessorRef = messageProcessor
        self.paymentsHelperRef = paymentsHelper
        self.paymentsCurrenciesRef = paymentsCurrencies
        self.paymentsEventsRef = paymentsEvents
        self.mobileCoinHelperRef = mobileCoinHelper
        self.spamChallengeResolverRef = spamChallengeResolver
        self.senderKeyStoreRef = senderKeyStore
        self.phoneNumberUtilRef = phoneNumberUtil
        self.webSocketFactoryRef = webSocketFactory
        self.legacyChangePhoneNumberRef = legacyChangePhoneNumber
        self.subscriptionManagerRef = subscriptionManager
        self.systemStoryManagerRef = systemStoryManager
        self.remoteMegaphoneFetcherRef = remoteMegaphoneFetcher
        self.sskJobQueuesRef = sskJobQueues
        self.contactDiscoveryManagerRef = contactDiscoveryManager
        self.callMessageHandlerRef = callMessageHandler
        self.notificationsManagerRef = notificationsManager
        self.messageSendLogRef = messageSendLog
    }

    public func signalProtocolStoreRef(for identity: OWSIdentity) -> SignalProtocolStore {
        switch identity {
        case .aci:
            return aciSignalProtocolStoreRef
        case .pni:
            return pniSignalProtocolStoreRef
        }
    }

    public static let warmCachesNotification = Notification.Name("WarmCachesNotification")

    public func warmCaches() {
        let warmCachesForObject: (String, () -> Void) -> Void = { name, action in
            InstrumentsMonitor.measure(category: "appstart", parent: "caches", name: name, block: action)
        }
        warmCachesForObject("signalProxy", SignalProxy.warmCaches)
        warmCachesForObject("newTSAccountManager", DependenciesBridge.shared.tsAccountManager.warmCaches)
        warmCachesForObject("fixLocalRecipient", fixLocalRecipientIfNeeded)
        warmCachesForObject("signalServiceAddressCache", signalServiceAddressCache.warmCaches)
        warmCachesForObject("signalService", signalService.warmCaches)
        warmCachesForObject("remoteConfigManager", remoteConfigManager.warmCaches)
        warmCachesForObject("blockingManager", blockingManager.warmCaches)
        warmCachesForObject("profileManager", profileManager.warmCaches)
        warmCachesForObject("receiptManager", receiptManager.prepareCachedValues)
        warmCachesForObject("OWSKeyBackupService", DependenciesBridge.shared.svr.warmCaches)
        warmCachesForObject("PinnedThreadManager", PinnedThreadManager.warmCaches)
        warmCachesForObject("typingIndicatorsImpl", typingIndicatorsImpl.warmCaches)
        warmCachesForObject("paymentsHelper", paymentsHelper.warmCaches)
        warmCachesForObject("paymentsCurrencies", paymentsCurrencies.warmCaches)
        warmCachesForObject("storyManager", StoryManager.setup)
        warmCachesForObject("deviceManager", DependenciesBridge.shared.deviceManager.warmCaches)
        warmCachesForObject("appExpiry") {
            DependenciesBridge.shared.db.read { tx in
                self.appExpiryRef.warmCaches(with: tx)
            }
        }

        NotificationCenter.default.post(name: SSKEnvironment.warmCachesNotification, object: nil)
    }

    /// Ensures the local SignalRecipient is correct.
    ///
    /// This primarily serves to ensure the local SignalRecipient has its own
    /// Pni (a one-time migration), but it also helps ensure that the value is
    /// always consistent with TSAccountManager's values.
    private func fixLocalRecipientIfNeeded() {
        databaseStorage.write { tx in
            guard let localIdentifiers = DependenciesBridge.shared.tsAccountManager.localIdentifiers(tx: tx.asV2Read) else {
                return  // Not registered yet.
            }
            guard let phoneNumber = E164(localIdentifiers.phoneNumber) else {
                return  // Registered with an invalid phone number.
            }
            let recipientMerger = DependenciesBridge.shared.recipientMerger
            _ = recipientMerger.applyMergeForLocalAccount(
                aci: localIdentifiers.aci,
                phoneNumber: phoneNumber,
                pni: localIdentifiers.pni,
                tx: tx.asV2Write
            )
        }
    }

    #if TESTABLE_BUILD

    public func setContactsManagerForUnitTests(_ contactsManager: ContactsManagerProtocol) {
        self.contactsManagerRef = contactsManager
    }

    public func setMessageSenderForUnitTests(_ messageSender: MessageSender) {
        self.messageSenderRef = messageSender
    }

    public func setNetworkManagerForUnitTests(_ networkManager: NetworkManager) {
        self.networkManagerRef = networkManager
    }

    public func setPaymentsHelperForUnitTests(_ paymentsHelper: PaymentsHelperSwift) {
        self.paymentsHelperRef = paymentsHelper
    }

    @objc
    public func setGroupsV2ForUnitTests(_ groupsV2: GroupsV2) {
        self.groupsV2Ref = groupsV2 as! GroupsV2Swift
    }

    #endif
}
